### ISSUE TITLE
Specs needed before any further work is done

### ISSUE THREAD

**Issue by mohawkjohn** [Arafatk/magick-rubyplot#4]
I've spent some time looking at your code, and I think that fundamentally you haven't done enough design.

I want you to pause your regular coding right now and switch over to writing specs.

Your first order of business is writing a spec helper that does image differencing in order to tell you if you've broken some feature in plotting. I want you to show me what you come up with before you continue coding.

CC @v0dro

**Comment by v0dro** [comment]
Works. I think the image differentiation spec will be very coupled to image magick so let it be in this repo.

However, I am of the opinion that common specs should be written in the rubyplot repo for better co-ordination: https://github.com/SciRuby/rubyplot/tree/master/spec

**Comment by mohawkjohn** [comment]
I think that’s fine, but I want him to consider all the possible plots he
needs to make as he continues with his coding. So if the specs need to go
in a different place, that’s fine, so long as he does them and starts
applying them.

**Comment by pgtgrly** [comment]
Greetings @mohawkjohn  @Arafatk  
@v0dro  and I just had a coding session where we discussed the Scripting layer based on the discussion[here](https://github.com/SciRuby/rubyplot/issues/1)

We came up with certain specs that based on the discussion to model the scripting layer.This might help you design your specs. 
```ruby
require 'spec_helper'

describe Rubyplot do
  before do
    @x1 = [1, 2, 3, 4, 5]
    @y1 = [10, 20, 30, 40, 50]
  end
  
  context "#line" do  
    it "creates a simple line graph" do
      a = Rubyplot.new
      a.line @x1, @y1 
      a.save "file_name.bmp"

      expect(equal_files("file_name.bmp", "line_graph.bmp")).to eq(true)
    end

    it "creates a line graph with points marked" do
      a = Rubyplot.new
      a.line @x1, @y1, points: true
      a.save "file_name.bmp"

      expect(equal_files("file_name.bmp", "line_points_graph.bmp")).to eq(true)
    end
  end

  context "#scatter" do
    it "creates a simple scatter graph" do
      a = Rubyplot.new
      a.scatter @x1, @y1
      a.save "file_name.bmp"

      expect(equal_files("file_name.bmp", "scatter_graph.bmp")).to eq(true)
    end
  end
end
```
```ruby
require 'spec_helper'

describe Rubyplot do
  before do
    @x1 = [-10, 0 , 5, 28] 
    @y1 = [1, 2, 3, 4]
    @x2 = [2, 4, 16]
    @y2 = [10, 20, -40]
  end

  context "#line #scatter" do
    it "creates a line and scatter graph" do
      a = Rubyplot.new
      a.line @x1, @y1
      a.scatter @x2, @y2
      a.save "file_name.bmp"

      expect(equal_files("file_name.bmp","line_scatter_graph.bmp")).to eq(true)
    end
  end
end
```
I have put them here so that they can be viewed even if the files are removed. Currently, They can be found [here](https://github.com/SciRuby/rubyplot/blob/master/spec/scripting_layer/single_plot_graph_spec.rb) and [here](https://github.com/SciRuby/rubyplot/blob/master/spec/scripting_layer/multi_plot_graph_spec.rb).

**Comment by mohawkjohn** [comment]
These are beautiful; excellent work.

I only have a few comments.

I like the name Rubyplot and think we should use this for the library. However, I don't think we should use this for the plots themselves. It would probably be better to stick to `Figure` or even `Axes` like in Python.

By the way, one of my most frequently used features in matplotlib is `fig.add_subplot(...)`, and I'd like you to think a bit about what creating multiple axes within a plot looks like in Rubyplot.

Can you also include an example which allows you to set the following basic attributes?
* Marker size
* Marker type
* Line size
* Line style
* Marker color
* Line color

Again, good work on this.

**Comment by pgtgrly** [comment]
@vodro @mohawkjohn I was just writing the mail about my today's progress and addressing this. I will iterate that over here first. Reading the comment decided to instantiate a class RbPlot inside RubyPlot which acts like Figure.
![gr-rubyplot](https://user-images.githubusercontent.com/17721748/41719420-98e1808c-757d-11e8-9bc7-4865a9004529.png)

I will hopefully include the example and specs for the mentioned basic attributes in the working prototype.

As for the Sub-Plots I am not sure if they can be implemented with GR Framework as this is a problem that has to dealt at and Artist Level and GR Framework abstracts that. I will however investigate GR framework/think of a bypass after I get the prototype working. I will be mailing on the ML soon.

**Comment by mohawkjohn** [comment]
Okay, but why call it RbPlot? This is confusing. Why not just call it Figure? Metaphors are really important, and people already understand Figure because they use Figure in Matplotlib and Matlab.

**Comment by Arafatk** [comment]
@pgtgrly @v0dro Thanks 
This helped a lot.

**Comment by mohawkjohn** [comment]
What thinking have you done regarding my comment from yesterday?

**Comment by Arafatk** [comment]
Hi John, 
Are you referring to me?

https://github.com/Arafatk/magick-rubyplot/commit/dad4455e0f6d797df9561231216e426f3f688ed5

**Comment by mohawkjohn** [comment]
Yes, but I was asking what thinking you'd done toward this comment:

https://github.com/Arafatk/magick-rubyplot/issues/4#issuecomment-399144280

Not what *work*, just what thinking.

I'll comment on this commit elsewhere.

**Comment by mohawkjohn** [comment]
I am referring to you. But I was actually asking about this comment: https://github.com/Arafatk/magick-rubyplot/issues/4#issuecomment-399144280

In addition, it's best practice — every time you do a commit that relates to an issue — to tag the issue number in the commit comment.

So your comment should be something like, "Adding Test mechanism to check image diffs against reference images. WIP #4." WIP stands for work in progress. When something fixes an issue, do "Fixes #4," and it'll actually automatically close the issue for you.

Now, with regards to https://github.com/Arafatk/magick-rubyplot/commit/dad4455e0f6d797df9561231216e426f3f688ed5, this is an overly complicated solution. Have you looked at the function matplotlib uses for all of its unit tests?

Here's the link:
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/testing/compare.py#L366

Why don't I like your solution? You're adding an additional gem requirement to a project. That makes everything an order of magnitude more complicated. You're occasionally going to need some additional gems (e.g., for magick), and that's unavoidable. But you should do everything you can to avoid introducing more complexity.

The meat of the method above is here:
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/testing/compare.py#L431

It's not hard to compute the RMS of two images. RMS is just the root-mean-square. DM me if you have trouble understanding the math.

**Comment by Arafatk** [comment]
All the potential extra plots to be added are described [here](https://docs.google.com/document/d/1PwlxewwKgw7yh-u1y5iJ2f3HPJ2htmlRXf_JCSE7Mj4/edit?usp=sharing) and the specs are in [this branch](https://github.com/Arafatk/magick-rubyplot/tree/extra-specs).
